File: PythonSubProg/tablemerge.py
def HatHunter(table):
    try:
        Hat = []
        for box in table:
            if box[1] == 1:
                curr_id = box[0]
                for i in range(2, len(box)):
                    header_value = box[i]
                    if header_value != None:
                        Hat.append([curr_id, header_value])
                        curr_id = id_cont(curr_id)
        return Hat
    except Exception as e:
        print(f"ERROR in HatHunter: {e}")
        print(f"Problematic data: {table}")
        return []  # Возвращаем пустой список вместо сбоя

def HatChecker(hat1, hat2):
    try:
        hat1 = HatHunter(hat1)
        hat2 = HatHunter(hat2)
        matches = []
        for i in range(len(hat1)):
            ha = hat1[i]
            a = ha[1]
            for x in range(len(hat2)):
                hb = hat2[x]
                b = hb[1]
                if a == b:
                    matches.append([ha[0], hb[0]])
        return matches
    except Exception as e:
        print(f"ERROR in HatChecker: {e}")
        return []  # Возвращаем пустой список вместо сбоя

#функция объединения двух таблиц на основе общих данных
def merge_tables(table1, table2):
    #проверка на общие элементы
    flag_common = HatChecker(table1, table2)
    if not flag_common:
        raise ValueError("There is no shared data")

    #проверка исключений
    exception_result = exception_process(table1, table2)
    if exception_result != None:
        return exception_result

    all_headers = select_headers(table1, table2) #сначала нужно получать словарь всех заголовков
    merged_table = create_newtable(table1, table2, all_headers) #потом уже создавать новую таблицу
    return merged_table
   
#функция для выделения всех заголовков в шапке в один словарь
def select_headers(table1, table2):
    headersdict = dict() 
    for table, table_name in [(table1, "Table 1"), (table2, "Table 2")]: #каждая итерация проходится по каждой таблице
        alltable_headers = HatHunter(table) 
        for header_id, header_name in alltable_headers: #теперь каждая итерация по каждому заголовку с его уникальным id
            #если заголовка ещё нет в словаре
            if header_name not in headersdict:
                headersdict[header_name] = {"table1_id": None, "table2_id": None} #пока None, ведь просто создаётся запись — потом None будут заполняться id заголовков с первой таблицы, а потом со второй
            #если такой заголовок уже есть
            if table_name == "Table 1":
                headersdict[header_name]["table1_id"] = header_id
            else:
                headersdict[header_name]["table2_id"] = header_id
    return headersdict #по итогу вернётся словарь, где явно видно, какие заголовки у таблиц общие, а какие нет
        
#функция создания новой мёрджнутой таблицы 
def create_newtable(table1, table2, all_uniq_headers): #принимает на вход словарь заголовков
    new_table = []

    #заполнение массивов уникальных заголовков
    uniq_headers1 = [] 
    uniq_headers2 = []
    common_headers = []
    for header, id in all_uniq_headers.items():
        if id["table1_id"] and id["table2_id"]: #в table1_id и в table2_id хранятся буквенные id
            common_headers.append(header)
        elif id["table1_id"] and not id["table2_id"]:
            uniq_headers1.append(header)
        elif not id["table1_id"] and id["table2_id"]:
            uniq_headers2.append(header)
    all_headers = common_headers + uniq_headers1 + uniq_headers2 #теперь создаётся один общий массив всех заголовков

    #составлене самой шапки
    curr_id = 'A'
    for header in all_headers:
        new_table.append([curr_id, 1, header]) #заполнение новой таблицы шапкой
        curr_id = id_cont(curr_id)

    #заполнение словарей заголовков с их id
    table1_id_header = dict()
    table2_id_header = dict()
    for cell in table1: 
        if cell[1] == 1:
            column_id = cell[0]
            for i in range(2, len(cell)): #индекс 2, потому что 0 — буквенный id ячейки, а 1 — числовой id ячейки
                header_name = cell[i]
                if header_name != None:
                    table1_id_header[header_name] = column_id
                    column_id = id_cont(column_id)
    for cell in table2: 
        if cell[1] == 1:
            column_id = cell[0]
            for i in range(2, len(cell)):
                header_name = cell[i]
                if header_name != None:
                    table2_id_header[header_name] = column_id
                    column_id = id_cont(column_id)  

    maxstroka_num_table1 = max(cell[1] for cell in table1) #максимальный номер строки в первой таблице
    maxstroka_num_table2 = max(cell[1] for cell in table2)

    #добавление строк из первой таблицы
    for stroka_num in range(2, maxstroka_num_table1 + 1): #1 итерация для каждой строки (по её номеру)
        curr_id = 'A'
        #нужно собрать все ячейки строки
        all_cells = dict()
        for cell in table1: #находим все ячейки первой строки первой таблицы
            if cell[1] == stroka_num:
                if len(cell) > 2:
                    all_cells[cell[0]] = cell[2] #установление номера текущей строки
                else:
                    all_cells[cell[0]] = ''
        for header in all_headers:
            cell_value = ''
            column_id = table1_id_header.get(header) #получаем значения у всех ключей заголовков в словаре
            if column_id in all_cells and column_id != None:
                cell_value = all_cells[column_id]
            new_table.append([curr_id, stroka_num, cell_value]) #теперь всё это формируется в строки и добавляется к уже составленной шапке в новой таблицы
            curr_id = id_cont(curr_id)

    #добавление строк из второй таблицы
    next_stroka_num = maxstroka_num_table1 + 1
    for stroka_num in range(2, maxstroka_num_table2 + 1): #находим все ячейки первой строки второй таблицы
        curr_id = 'A'
        #нужно собрать все ячейки строки
        all_cells = dict()
        for cell in table2: #находим все ячейки первой строки первой таблицы
            if cell[1] == stroka_num:
                if len(cell) > 2:
                    all_cells[cell[0]] = cell[2]
                else:
                    all_cells[cell[0]] = ''
        for header in all_headers:
            cell_value = ''
            column_id = table2_id_header.get(header) #получаем значения у всех ключей заголовков в словаре
            if column_id in all_cells and column_id != None:
                cell_value = all_cells[column_id]
            new_table.append([curr_id, stroka_num, cell_value]) #теперь всё это формируется в строки и добавляется к уже составленной шапке в новой таблицы
            curr_id = id_cont(curr_id)
        next_stroka_num += 1
    return new_table


#функция для продолжения нумерации ячеек 
def id_cont(cell_id):
    #если ячейка пуста — начинаем с первой ячейки A
    if cell_id == None:
        return 'A'
    '''для определения правильного двузначного и большеидентификатора будет строиться 
    следующая логика: текущий буквенный идентификатор сначала превратится в число, 
    потом это число увеличится и вернётся новый правильный буквенный идентификатор'''
    def id_chislo(cell_id): 
        chislo = 0
        for letter in cell_id: #для каждой буквы в id
            chislo = chislo * 26 + (ord(letter.upper()) - ord('A') + 1) #вычисление номера буквы по латинскому алфавиту
        return chislo

    def chislo_newid(chislo):
        cell_id = ''    
        while chislo > 0:
            ostatok = (chislo - 1) % 26
            chislo = (chislo - 1) // 26
            cell_id = chr(ord('A') + ostatok) + cell_id 
        return cell_id
    curr_chislo = id_chislo(cell_id)
    next_chislo = curr_chislo + 1 #номер следующей ячейки от заданной
    return chislo_newid(next_chislo) #id этой следующей
    

#функция для переиндексации ячеек в таблице
def reindex_id(table, start_cell_id):
    if table == None:
        return []
    reindex_table = [] 
    curr_cell_id = start_cell_id
    for cell in table:
        new_cell = [curr_cell_id] + cell[1:] #заменяется буква у ячейки с сохранением всех остальных данных
        reindex_table.append(new_cell)
        curr_cell_id = id_cont(curr_cell_id)
    return reindex_table


#функция для нахождения последней ячейки (по букве)
def get_maxid(table):
    if table == None:
        return None
    #для правильного определения большего идентификатора
    def id_chislo(id_stroka):
        chislo = 0
        for index, letter in enumerate(reversed(id_stroka)): #перебор по каждому индексу (для веса разряда) и самой букве
            chislo += (ord(letter.upper()) - ord('A') + 1) * (26 ** index) #вычисление номера буквы по латинскому алфавиту и умножение на вес разряда
        return chislo #значение идентификатора (по которому будет сравниваться)
    all_id = [cell[0] for cell in table] #все id ячеек (без чисел)
    maxid_stroka = all_id[0] 
    maxid_chislo = id_chislo(maxid_stroka) 
    for cell_id in all_id[1:]: #начнётся перебор со второго id (так как первый уже обозначен, как максимальный) и будет переобозначаться максимальный id
        cell_value = id_chislo(cell_id)
        if cell_value > maxid_chislo: #переобозначаться максимальный id будет просто в сравнении числовых значений id
            maxid_stroka = cell_id
            maxid_chislo = cell_value
    return maxid_stroka


#функция для обработки исключений
def exception_process(table1, table2):
    if table1 == None and table2 == None:
        raise ValueError("Both tables are None")
        return None
    if table1 != None and table2 == None:
        print("Table #2 is None. Return table #1")
        return table1
    if table1 == None and table2 != None:
        print("Table #1 is None. Return table #2")
        return table2

File: ExcelDBconsole/TableMerge.cs
using System;
using System.Collections.Generic;
using System.Linq;
namespace ExcelBDConsol
{
    public static class TableMerge
    {
        //главная функция объединения двух таблиц на основе общих данных
        public static List<List<object>> MergeTables(List<List<object>> table1, List<List<object>> table2)
        {
            //проверка на общие элементы
            var flagCommon = HatChecker(table1, table2);
            if (flagCommon.Count == 0)
            {
                throw new ArgumentException("There is no shared data");
            }
            //проверка исключений
            var exceptionResult = ExceptionProcess(table1, table2);
            if (exceptionResult != null)
            {
                return exceptionResult;
            }
            var all_headers = SelectHeaders(table1, table2); //сначала нужно получать словарь всех заголовков
            var mergedTable = CreateNewTable(table1, table2, all_headers); //потом уже создавать новую таблицу
            return mergedTable;
        }

        //функция для выделения шапки
        private static List<Tuple<string, string>> HatHunter(List<List<object>> table)
        {
            var hat = new List<Tuple<string, string>>();

            foreach (var box in table)
            {
                if (box.Count > 1 && box[1] is int && (int)box[1] == 1)
                {
                    string currId = box[0]?.ToString() ?? "A";
                    for (int i = 2; i < box.Count; i++)
                    {
                        var headerValue = box[i]?.ToString();
                        if (!string.IsNullOrEmpty(headerValue))
                        {
                            hat.Add(new Tuple<string, string>(currId, headerValue));
                            currId = IdCont(currId);
                        }
                    }
                }
            }
            return hat;
        }

        //функция для сравнения шапок
        private static List<Tuple<string, string>> HatChecker(List<List<object>> table1, List<List<object>> table2)
        {
            var hat1 = HatHunter(table1);
            var hat2 = HatHunter(table2);
            var matches = new List<Tuple<string, string>>();
            foreach (var ha in hat1)
            {
                string a = ha.Item2;
                foreach (var hb in hat2)
                {
                    string b = hb.Item2;
                    if (a == b)
                    {
                        matches.Add(new Tuple<string, string>(ha.Item1, hb.Item1));
                    }
                }
            }
            return matches;
        }

        //ВЫДЕЛЕНИЕ ЗАГОЛОВКОВ
        //функция для выделения всех заголовков в шапке в один словарь
        private static Dictionary<string, HeaderInfo> SelectHeaders(List<List<object>> table1, List<List<object>> table2)
        {
            var headersDict = new Dictionary<string, HeaderInfo>();
            ProcessTableHeaders(table1, headersDict, "Table1");
            ProcessTableHeaders(table2, headersDict, "Table2");
            return headersDict;
        }
        //функция для обработки таблицы
        private static void ProcessTableHeaders(List<List<object>> table, Dictionary<string, HeaderInfo> headersDict, string tableName)
        {
            var tableHeaders = HatHunter(table);
            foreach (var header in tableHeaders)
            {
                string headerName = header.Item2;
                string headerId = header.Item1;
                if (!headersDict.ContainsKey(headerName))
                {
                    headersDict[headerName] = new HeaderInfo();
                }
                if (tableName == "Table1")
                {
                    headersDict[headerName].Table1Id = headerId;
                }
                else
                {
                    headersDict[headerName].Table2Id = headerId;
                }
            }
        }
        //класс для хранения заголовков и значений
        private class HeaderInfo
        {
            public string Table1Id { get; set; }
            public string Table2Id { get; set; }
        }

        //СОЗДАНИЕ ОБЪЕДИНЁННОЙ ТАБЛИЦЫ
        //функция создания новой мёрджнутой таблицы 
        private static List<List<object>> CreateNewTable(List<List<object>> table1, List<List<object>> table2, Dictionary<string, HeaderInfo> allUniqHeaders)
        {
            var newTable = new List<List<object>>();
            //заполнение массивов уникальных заголовков
            var uniqueHeaders1 = new List<string>();
            var uniqueHeaders2 = new List<string>();
            var commonHeaders = new List<string>();
            foreach (var kvp in allUniqHeaders)
            {
                string header = kvp.Key;
                var info = kvp.Value;
                if (!string.IsNullOrEmpty(info.Table1Id) && !string.IsNullOrEmpty(info.Table2Id))
                {
                    commonHeaders.Add(header);
                }
                else if (!string.IsNullOrEmpty(info.Table1Id) && string.IsNullOrEmpty(info.Table2Id))
                {
                    uniqueHeaders1.Add(header);
                }
                else if (string.IsNullOrEmpty(info.Table1Id) && !string.IsNullOrEmpty(info.Table2Id))
                {
                    uniqueHeaders2.Add(header);
                }
            }
            //заполнение массивов уникальных заголовков
            var allHeaders = new List<string>();
            allHeaders.AddRange(commonHeaders);
            allHeaders.AddRange(uniqueHeaders1);
            allHeaders.AddRange(uniqueHeaders2);
            //составление шапки таблицы
            string currId = "A";
            foreach (var header in allHeaders)
            {
                newTable.Add(new List<object> { currId, 1, header });
                currId = IdCont(currId);
            }
            //заполнение словарей заголовков с их ID
            var table1IdHeader = BuildHeaderIdDictionary(table1);
            var table2IdHeader = BuildHeaderIdDictionary(table2);
            //максимальные номера строк
            int maxRowNumTable1 = GetMaxRowNumber(table1);
            int maxRowNumTable2 = GetMaxRowNumber(table2);
            //добавление строк из первой таблицы
            for (int rowNum = 2; rowNum <= maxRowNumTable1; rowNum++)
            {
                currId = "A";
                var allCells = GetRowCells(table1, rowNum);
                foreach (var header in allHeaders)
                {
                    string cellValue = "";
                    string columnId = table1IdHeader.ContainsKey(header) ? table1IdHeader[header] : null;

                    if (columnId != null && allCells.ContainsKey(columnId))
                    {
                        cellValue = allCells[columnId]?.ToString() ?? "";
                    }
                    newTable.Add(new List<object> { currId, rowNum, cellValue });
                    currId = IdCont(currId);
                }
            }
            //добавление строк из второй таблицы
            int nextRowNum = maxRowNumTable1 + 1;
            for (int rowNum = 2; rowNum <= maxRowNumTable2; rowNum++)
            {
                currId = "A";
                var allCells = GetRowCells(table2, rowNum);
                foreach (var header in allHeaders)
                {
                    string cellValue = "";
                    string columnId = table2IdHeader.ContainsKey(header) ? table2IdHeader[header] : null;
                    if (columnId != null && allCells.ContainsKey(columnId))
                    {
                        cellValue = allCells[columnId]?.ToString() ?? "";
                    }

                    newTable.Add(new List<object> { currId, nextRowNum, cellValue });
                    currId = IdCont(currId);
                }
                nextRowNum++;
            }
            return newTable;
        }
        //функция для построения словарей заголовков
        private static Dictionary<string, string> BuildHeaderIdDictionary(List<List<object>> table)
        {
            var dict = new Dictionary<string, string>();
            foreach (var cell in table)
            {
                if (cell.Count > 1 && cell[1] is int && (int)cell[1] == 1)
                {
                    string columnId = cell[0]?.ToString();

                    for (int i = 2; i < cell.Count; i++)
                    {
                        string headerName = cell[i]?.ToString();
                        if (!string.IsNullOrEmpty(headerName))
                        {
                            dict[headerName] = columnId;
                            columnId = IdCont(columnId);
                        }
                    }
                }
            }
            return dict;
        }
        //функция для выделения максимального номера строки
        private static int GetMaxRowNumber(List<List<object>> table)
        {
            int max = 0;
            foreach (var cell in table)
            {
                if (cell.Count > 1 && cell[1] is int)
                {
                    int rowNum = (int)cell[1];
                    if (rowNum > max)
                    {
                        max = rowNum;
                    }
                }
            }
            return max;
        }
        //функция для сбора всех ячеек строки в словарь
        private static Dictionary<string, object> GetRowCells(List<List<object>> table, int rowNum)
        {
            var cells = new Dictionary<string, object>();

            foreach (var cell in table)
            {
                if (cell.Count > 1 && cell[1] is int && (int)cell[1] == rowNum) //нахождение всех ячеек определённой строки
                {
                    string columnId = cell[0]?.ToString();
                    object value = cell.Count > 2 ? cell[2] : "";
                    cells[columnId] = value;
                }
            }
            return cells;
        }

        //функция для продолжения нумерации ячеек 
        private static string IdCont(string cellId)
        {
            //если ячейка пуста — начинаем с первой ячейки A
            if (string.IsNullOrEmpty(cellId))
            {
                return "A";
            }
            /*для определения правильного двузначного и больше идентификатора будет строиться 
            следующая логика: текущий буквенный идентификатор сначала превратится в число, 
            потом это число увеличится и вернётся новый правильный буквенный идентификатор*/
            //превращение буквенного id в число
            int NumberFromId(string id)
            {
                int number = 0;
                foreach (char letter in id.ToUpper())
                {
                    number = number * 26 + (letter - 'A' + 1); //вычисление номера буквы по латинскому алфавиту
                }
                return number;
            }
            //превращение числового id обратно в буквенный
            string IdFromNumber(int number)
            {
                string id = "";
                while (number > 0)
                {
                    int remainder = (number - 1) % 26;
                    number = (number - 1) / 26;
                    id = (char)('A' + remainder) + id;
                }
                return id;
            }
            int currentNumber = NumberFromId(cellId);
            int nextNumber = currentNumber + 1; //номер следующей ячейки от заданной
            return IdFromNumber(nextNumber); //id этой следующей
        }

        //функция для обработки исключений
        private static List<List<object>> ExceptionProcess(List<List<object>> table1, List<List<object>> table2)
        {
            if (table1 == null && table2 == null)
            {
                throw new ArgumentException("Both tables are None");
            }
            if (table1 != null && table2 == null)
            {
                Console.WriteLine("Table #2 is None. Return table #1");
                return table1;
            }
            if (table1 == null && table2 != null)
            {
                Console.WriteLine("Table #1 is None. Return table #2");
                return table2;
            }
            return null; //нет исключительной ситуации
        }
    }
}